File: lecture_010(RecursionwithArraylists)/l_006_printsubsequence.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

void printsubsequence(string ques, string ans) {
    if(ques.length()==0) {
        cout<<ans<<endl;
        return;
    }
    char ch = ques[0];
    string roq = ques.substr(1);

    printsubsequence(roq,ans+"-");
    printsubsequence(roq,ans+ch);
}

int main() {
    string str;
    getline(cin,str);
    printsubsequence(str,"");

    return 0;
}

File: lecture_010(RecursionwithArraylists)/l_006_printsubsequence.java
import java.util.Scanner;
public class l_006_printsubsequence {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        String str = scn.nextLine();
        printSubsequence(str,"");
    }
    public static void printSubsequence(String ques, String ans) {
        if(ques.length()==0) {
            System.out.println(ans);
            return;
        }
        char ch = ques.charAt(0);
        String roq = ques.substring(1);

        printSubsequence(roq,ans+"-");     //wrong 
        printSubsequence(roq,ans+ch);      //right
    }
}